query/src/lib/mutation.ts

```diff
@@ -0,0 +1,113 @@
+import { inject, Injectable, InjectionToken, Signal } from '@angular/core';
+import { injectQueryClient } from './query-client';
+import {
+  DefaultError,
+  MutationObserver,
+  MutationObserverOptions,
+  MutationObserverResult,
+  notifyManager,
+} from '@tanstack/query-core';
+import { firstValueFrom, isObservable, Observable, shareReplay } from 'rxjs';
+import { toSignal } from '@angular/core/rxjs-interop';
+
+type CreateMutationOptions<
+  TData = unknown,
+  TError = DefaultError,
+  TVariables = void,
+  TContext = unknown
+> = Omit<
+  MutationObserverOptions<TData, TError, TVariables, TContext>,
+  'mutationFn'
+> & {
+  mutationFn: (variables: TVariables) => Promise<TData> | Observable<TData>;
+};
+
+type MutationResult<
+  TData = unknown,
+  TError = DefaultError,
+  TVariables = void,
+  TContext = unknown
+> = {
+  mutate: MutationObserver<TData, TError, TVariables, TContext>['mutate'];
+  reset: MutationObserver<TData, TError, TVariables, TContext>['reset'];
+  setOptions: MutationObserver<
+    TData,
+    TError,
+    TVariables,
+    TContext
+  >['setOptions'];
+  result$: Observable<
+    MutationObserverResult<TData, TError, TVariables, TContext>
+  >;
+  toSignal: () => Signal<MutationObserverResult<TData, TError, TVariables, TContext> | undefined>
+};
+
+@Injectable({ providedIn: 'root' })
+class Mutation {
+  private instance = injectQueryClient();
+
+  use<
+    TData = unknown,
+    TError = DefaultError,
+    TVariables = unknown,
+    TContext = unknown
+  >(
+    options: CreateMutationOptions<TData, TError, TVariables, TContext>
+  ): MutationResult<TData, TError, TVariables, TContext> {
+    const mutationObserver = new MutationObserver<
+      TData,
+      TError,
+      TVariables,
+      TContext
+    >(this.instance, {
+      ...options,
+      mutationFn: (variables: TVariables): Promise<TData> => {
+        const result: Promise<TData> | Observable<TData> =
+          options.mutationFn(variables);
+        if (isObservable(result)) return firstValueFrom(result);
```

Comment:
If you pass an observable the return type is typed in the mutation result?

Reply:
Yes it gets passed through to the result.

![grafik](https://github.com/ngneat/query/assets/8077014/db1d77ce-5f0b-4878-8434-88b9300e37d8)

Reply:
When you subscribe manually you get the Todo type in the data?

Reply:
yes, you also get the type pulled through into the `next` emission of the manual subscription.

![grafik](https://github.com/ngneat/query/assets/8077014/3a692f16-245a-4fd0-9ba5-aa126114fae9)

Is there anything wrong with the typing for you? For me atleast Webstorm cant infer the typings in the template but i think this is a Webstorm problem, since it works in VS Code with no problem.

Reply:
I didn't notice this `mutationFn: (variables: TVariables) => Promise<TData> | Observable<TData>`.